File: MyNote_OC_Online/MyNote_OC_Online/MasterViewController.m
//
//  MasterViewController.m
//  MyNote_OC_Online
//
//  Created by 俊杰李 on 2020/5/18.
//  Copyright © 2020 俊杰李. All rights reserved.
//

#import "MasterViewController.h"
#import "DetailViewController.h"

#import "HttpUtil.h"
@interface MasterViewController ()
//保存数据列表
@property(nonatomic, strong) NSMutableArray *notes;
@end

@implementation MasterViewController
NSString* myAccount= @"l536510961@126.com";
- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.navigationItem.leftBarButtonItem = self.editButtonItem;
    
    self.detailViewController = (DetailViewController *)[[self.splitViewController.viewControllers lastObject] topViewController];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(reload:) name:@"reload" object:nil];
    
    [self loadAllNotes];
}

- (void)viewWillAppear:(BOOL)animated {
    self.clearsSelectionOnViewWillAppear = self.splitViewController.isCollapsed;
    [super viewWillAppear:animated];
}


- (void)insertNewObject:(id)sender {
    if (!self.notes) {
        self.notes = [[NSMutableArray alloc] init];
    }
    [self.notes insertObject:[NSDate date] atIndex:0];
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:0 inSection:0];
    [self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
}


#pragma mark - Segues

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([[segue identifier] isEqualToString:@"showDetail"]) {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        NSDictionary *object = self.notes[indexPath.row];
        DetailViewController *controller = (DetailViewController *)[[segue destinationViewController] topViewController];
        controller.detailItem = object;
        controller.navigationItem.leftBarButtonItem = self.splitViewController.displayModeButtonItem;
        controller.navigationItem.leftItemsSupplementBackButton = YES;
        self.detailViewController = controller;
    }
}


#pragma mark - Table View

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.notes.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    
    NSDictionary *object = self.notes[indexPath.row];
    
    cell.textLabel.text = object[@"Content"];
    cell.detailTextLabel.text = object[@"CDate"];
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        NSDictionary* note = self.notes[indexPath.row];
        [[HttpUtil alloc] removeById:myAccount mid:note[@"ID"] completionHandler:^(NSDictionary *resDict, NSError *error) {
            if(error){
                [self showMessage:error.description];
            }else{
//                [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
//
                [self loadAllNotes];
//                [self.notes removeObjectAtIndex:indexPath.row];
            }
        }];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view.
    }
}


-(void)reload:(NSNotification*) notification{
    [self loadAllNotes];
}

-(void) loadAllNotes{
    [[HttpUtil alloc] loadAll:myAccount completionHandler:^(NSDictionary *resDict, NSError *error) {
        if(error){
            [self showMessage: error.description];
            return;
        }
        NSNumber *resultCode = resDict[@"ResultCode"];
        if ([resultCode integerValue] >= 0) {
            self.notes = resDict[@"Record"];
            [self.tableView reloadData];
        } else {
            NSString *errorStr = [resultCode errorMessage];
            [self showMessage: errorStr];
        }
    }];
}
-(void) showMessage:(NSString*) message{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"错误信息" message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alertController addAction:okAction];
    //显示
    [self presentViewController:alertController animated:true completion:nil];
}
@end

File: MyNote_Swift/MyNote_Swift/MasterViewController.swift
//
//  MasterViewController.swift
//  MyNote_Swift
//
//  Created by 俊杰李 on 2020/5/13.
//  Copyright © 2020 俊杰李. All rights reserved.
//

import UIKit

class MasterViewController: UITableViewController {

    var detailViewController: DetailViewController? = nil
    var notes = NSMutableArray()


    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        navigationItem.leftBarButtonItem = editButtonItem

        if let split = splitViewController {
            let controllers = split.viewControllers
            detailViewController = (controllers[controllers.count-1] as! UINavigationController).topViewController as? DetailViewController
        }
        
        let dao:NoteDao = NoteDao.sharedInstance
        notes = dao.quaryAll()
        
        NotificationCenter.default.addObserver(self, selector: #selector(reload(_:)), name: Notification.Name(rawValue: "reaload"), object: nil)
         
    }

      // MARK: --处理通知
    @objc func reload(_ notification:Notification){
        let resList = notification.object as! NSMutableArray
        self.notes = resList
        self.tableView.reloadData()
    }
    
    
    override func viewWillAppear(_ animated: Bool) {
        clearsSelectionOnViewWillAppear = splitViewController!.isCollapsed
        super.viewWillAppear(animated)
    }

    // MARK: - Segues
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "showDetail" {
            if let indexPath = tableView.indexPathForSelectedRow {
                let note = notes[indexPath.row] as! Note
                let controller = (segue.destination as! UINavigationController).topViewController as! DetailViewController
                controller.detailItem = note
                controller.navigationItem.leftBarButtonItem = splitViewController?.displayModeButtonItem
                controller.navigationItem.leftItemsSupplementBackButton = true
                detailViewController = controller
            }
        }
    }

    // MARK: - Table View

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.notes.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        let note = self.notes[indexPath.row] as! Note
        cell.textLabel!.text = note.content
        let date = note.date.description
        cell.detailTextLabel!.text = date
        return cell
    }

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        // Return false if you do not want the specified item to be editable.
        return true
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            let dao = NoteDao.sharedInstance
            dao.delete(note: notes[indexPath.row] as! Note)
            notes.removeObject(at: indexPath.row)
            tableView.deleteRows(at: [indexPath], with: .fade)
        } else if editingStyle == .insert {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view.
        }
    }


}

